src/gerbil/runtime/mop.ss

```diff
@@ -6,449 +6,554 @@ package: gerbil/runtime
+   ((find type-final? direct-supers)
+    => (cut error "Cannot extend final class" <>)))
+
+  (let* ((struct-super (base-struct/list direct-supers)) ;; super struct, if any
+         (precedence-list (class-linearize-mixins direct-supers)) ;; class precedence list, excluding current class
+         ((values all-slots slot-table)
+          (compute-class-slots struct-super precedence-list direct-slots))
+         (alist
+          [[direct-slots: . direct-slots]
+           [direct-supers: . direct-supers]
+           alist ...])
+         (constructor* (or constructor (find-super-constructor direct-supers))))
+
+    (make-type-descriptor* id name struct-super
+                           precedence-list all-slots alist
+                           constructor* slot-table #f)))
+
+(def (class-precedence-list klass)
+  (cons klass (type-descriptor-precedence-list klass)))
+
+(def (struct-precedence-list strukt)
+  (cons strukt (cond ((##type-super strukt) => struct-precedence-list) (else []))))
 
-  (K lst '()))
+(def (class-linearize-mixins klass-lst)
+  (c3-linearize [] klass-lst class-precedence-list eqv? ##type-name))
 
 (def (make-class-predicate klass)
-  (if (assgetq final: (type-descriptor-plist klass))
-    (lambda (obj)
-      (direct-class-instance? klass obj))
-    (lambda (obj)
-      (class-instance? klass obj))))
+  (if (type-final? klass)
+    (cut direct-class-instance? klass <>)
+    (cut class-instance? klass <>)))
+
+(def (if-class-slot-field klass slot if-struct if-struct-field if-class-slot)
+  ;;(DBG if-class-slot-field: (type-id klass) slot if-struct if-struct-slot if-class-slot)
+  (let (field (hash-get (type-descriptor-slot-table klass) slot))
+    #;(DBG if-class-slot-field-2: field (type-final? klass) (type-struct? klass)
+         'base-struct (type-id (base-struct/1 klass))
+         'vector-length (let (strukt (base-struct/1 klass))
+                          (and strukt (vector-length (type-descriptor-all-slots strukt)))) 'result:)
+    (cond
+     ((or (type-final? klass) (type-struct? klass))
+      (if-struct klass field))
+     ((let (strukt (base-struct/1 klass))
+        (and strukt (< field (vector-length (type-descriptor-all-slots strukt)))))
+      (if-struct-field klass field))
+     (else
+      (if-class-slot klass slot field)))));;)
 
 (def (make-class-slot-accessor klass slot)
+  ;;(DBG make-class-slot-accessor: (type-id klass) slot ':
+  (if-class-slot-field klass slot
+    make-struct-field-accessor*
+    make-struct-subclass-field-accessor
+    make-class-cached-slot-accessor));;)
+
+(def (make-struct-subclass-field-accessor klass field)
   (lambda (obj)
-    (slot-ref obj slot)))
+    (if (class-instance? klass obj)
+      (unchecked-field-ref obj field)
+      (error "Trying to get a slot of an object that is not a class instance"
+        (vector-ref (type-descriptor-all-slots klass) field) obj klass))))
+
+(def (make-class-cached-slot-accessor klass slot field)
```

Comment:
what is this for?

Reply:
What is what for?

Reply:
this method!

Reply:
`make-struct-subclass-field-accessor` or `make-class-cached-slot-accessor` ? Either way, they are both cases used by `make-class-slot-accessor`.

Reply:
ok.